File: libgdx/pt_octree_libgdx-android/src/com/amaisu/ptoctreelibgdx/MCTemplate.java
package com.amaisu.ptoctreelibgdx;

public class MCTemplate {
	public Triangle triangles_[];
	public int n_triangles_;
	public int index_; // See (Lorensen, 1987)
	public boolean complement_;

	public MCTemplate(final int n_triangles, final int index) {
		triangles_ = new Triangle[n_triangles];
		for (int i = 0; i < n_triangles; i++) {
			triangles_[i] = new Triangle();
		}
		this.index_ = index;
		this.n_triangles_ = n_triangles;
		complement_ = false;
	}

	public void Render() {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].Render(0, 1);
		}
	}

	public void CopyTriangles(final MCTemplate source) {
		if (source.n_triangles_ != n_triangles_) {
			return;
		}
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].normal_.x_ = source.triangles_[i].normal_.x_;
			triangles_[i].normal_.y_ = source.triangles_[i].normal_.y_;
			triangles_[i].normal_.z_ = source.triangles_[i].normal_.z_;
			for (int j = 0; j < 3; j++) {
				triangles_[i].vertices_[j].x_ = source.triangles_[i].vertices_[j].x_;
				triangles_[i].vertices_[j].y_ = source.triangles_[i].vertices_[j].y_;
				triangles_[i].vertices_[j].z_ = source.triangles_[i].vertices_[j].z_;
			}
		}
	}

	public MCTemplate RotateX(final double degrees) {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].RotateX(degrees);
		}
		return this;
	}

	public MCTemplate RotateY(final double degrees) {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].RotateY(degrees);
		}
		return this;
	}

	public MCTemplate RotateZ(final double degrees) {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].RotateZ(degrees);
		}
		return this;
	}

	public MCTemplate MirrorX() {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].MirrorX();
		}
		return this;
	}

	public MCTemplate MirrorY() {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].MirrorY();
		}
		return this;
	}

	public MCTemplate MirrorZ() {
		for (int i = 0; i < n_triangles_; i++) {
			triangles_[i].MirrorZ();
		}
		return this;
	}

	public void set_complement(final boolean complement) {
		if (complement_ != complement) {
			for (int i = 0; i < n_triangles_; i++) {
				Vector3 v = new Vector3();

				v.x_ = triangles_[i].vertices_[1].x_;
				v.y_ = triangles_[i].vertices_[1].y_;
				v.z_ = triangles_[i].vertices_[1].z_;

				triangles_[i].vertices_[1].x_ = triangles_[i].vertices_[2].x_;
				triangles_[i].vertices_[1].y_ = triangles_[i].vertices_[2].y_;
				triangles_[i].vertices_[1].z_ = triangles_[i].vertices_[2].z_;

				triangles_[i].vertices_[2].x_ = v.x_;
				triangles_[i].vertices_[2].y_ = v.y_;
				triangles_[i].vertices_[2].z_ = v.z_;

				triangles_[i].normal_.MirrorX().MirrorY().MirrorZ();
			}
		}
		complement_ = complement;
	}
}

File: furai/nacl/MCTemplate.cpp
/*
 * Copyright (c) 2012 Allan Yoshio Hasegawa
 * 
 * Permission is hereby granted, free of charge, to any
 * person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the
 * Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute,
 * sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall
 * be included in all copies or substantial portions of the
 * Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
 * KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
 * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
 * OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include <MCTemplate.h>

MCTemplate::MCTemplate(const int n_triangles, const int index)
    : n_triangles_(n_triangles),
      index_(index),
      complement_(false) {
  triangles_ = new Triangle[n_triangles];
}

MCTemplate::~MCTemplate() {
  delete[] triangles_;
}

void MCTemplate::Render() {
  for (int i = 0; i < n_triangles_; i++) {
    //triangles_[i].Render();
  }
}

void MCTemplate::CopyTriangles(const MCTemplate& source) {
  if (source.n_triangles_ != n_triangles_) {
    return;
  }
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i] = source.triangles_[i];
  }
}

MCTemplate* MCTemplate::RotateX(const double degrees) {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].RotateX(degrees);
  }
  return this;
}

MCTemplate* MCTemplate::RotateY(const double degrees) {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].RotateY(degrees);
  }
  return this;
}

MCTemplate* MCTemplate::RotateZ(const double degrees) {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].RotateZ(degrees);
  }
  return this;
}

MCTemplate* MCTemplate::MirrorX() {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].MirrorX();
  }
  return this;
}

MCTemplate* MCTemplate::MirrorY() {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].MirrorY();
  }
  return this;
}

MCTemplate* MCTemplate::MirrorZ() {
  for (int i = 0; i < n_triangles_; i++) {
    triangles_[i].MirrorZ();
  }
  return this;
}

void MCTemplate::set_complement(const bool complement) {
  if (complement_ != complement) {
    for (int i = 0; i < n_triangles_; i++) {
      triangles_[i].normal_.MirrorX()->MirrorY()->MirrorZ();
    }
  }
  complement_ = complement;
}
